### PR TITLE
Add guard override capability for tensor subclass metadata

### PR DESCRIPTION
**Pull request by mlazos**
Fixes https://github.com/pytorch/pytorch/issues/114405

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130780
* #130779

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

**Review by anijain2305 — APPROVED** [review]

File: test/dynamo/test_subclasses.py [hunk 1]
```diff
@@ -1304,74 +1372,53 @@ def fn(w):
...
+        _check_recompiles(self, lambda x: x * x, (x,), (x2,), False)
+        _check_recompiles(self, lambda x: x * x, (x,), (x3,), True)
 
-            def __repr__(self):
-                a_repr = repr(self.a)
-                return f"SubclassTensor({a_repr})"
+    def test_tensor_subclass_ctx_custom_guards_override(self):
+        class CtxSubclassTensorCustomGuardFn(CtxSubclassTensor):
+            @classmethod
+            def __metadata_guard__(cls, orig_data, other):
```
Comment (bdhirsh) [inline test/dynamo/test_subclasses.py:1384]: Now that we're guarding properly, we probably also want to add this guard function to `DTensor` ([link](https://github.com/pytorch/pytorch/blob/main/torch/distributed/_tensor/api.py#L208)) or else we might see DTensor get a little slower with compile when graphs are small (DTensor has a lot of metadata, but in practice I don't think it wants to guard on all of it).

@wanchaol is it fair to say we probably dont' want to guard on the entire `DTensor._spec` (which contains a bunch of size/stride metadata that is redundant)? I would guess that we want the guard to at the very least only assert on `_spec.device_mesh` and `_spec.placements`,

```

**Review by bdhirsh — APPROVED** [review]

**Comment by wanchaol** [comment]
Nice to see this is enabled!

**Pull request closed without merging**